ruma-api-macros/src/derive_outgoing.rs

```diff
@@ -0,0 +1,184 @@
+    let derive_deserialize = if no_deserialize_in_attrs(&input.attrs) {
+        TokenStream::new()
+    } else {
+        quote!(::ruma_api::exports::serde::Deserialize)
+    };
+
+    let (mut fields, struct_kind): (Vec<_>, _) = match input.data {
+        Data::Enum(_) | Data::Union(_) => {
+            panic!("#[derive(Outgoing)] is only supported for structs")
+        }
+        Data::Struct(s) => match s.fields {
+            Fields::Named(fs) => {
+                (fs.named.into_pairs().map(Pair::into_value).collect(), StructKind::Struct)
+            }
+            Fields::Unnamed(fs) => {
+                (fs.unnamed.into_pairs().map(Pair::into_value).collect(), StructKind::Tuple)
+            }
+            Fields::Unit => return Ok(impl_outgoing_with_incoming_self(input.ident)),
+        },
+    };
+
+    let mut any_attribute = false;
+
+    for field in &mut fields {
+        let mut found_wrap_incoming = false;
+
+        let mut remaining_attrs = Vec::new();
+        for attr in mem::replace(&mut field.attrs, Vec::new()) {
+            if attr.path.is_ident("wrap_incoming") {
+                if found_wrap_incoming {
+                    return Err(syn::Error::new_spanned(
+                        attr,
+                        "duplicate #[wrap_incoming] attribute",
+                    ));
+                }
+                found_wrap_incoming = true;
+                any_attribute = true;
+            } else {
+                remaining_attrs.push(attr);
+            }
+        }
+        field.attrs = remaining_attrs;
+
+        if found_wrap_incoming {
+            strip_lifetimes(field);
+        }
+    }
+
+    if !any_attribute {
+        return Ok(impl_outgoing_with_incoming_self(input.ident));
+    }
+
+    let original_ident = &input.ident;
+    let (original_impl_gen, original_ty_gen, _) = input.generics.split_for_impl();
+
+    let vis = input.vis;
+    let doc = format!("'Incoming' variant of [{ty}](struct.{ty}.html).", ty = &input.ident);
+    let incoming_ident = format_ident!("Incoming{}", original_ident, span = Span::call_site());
+    let mut gen_copy = input.generics.clone();
+    let (impl_gen, ty_gen) = split_for_impl_lifetime_less(&mut gen_copy);
+
+    let struct_def = match struct_kind {
+        StructKind::Struct => quote! { { #(#fields,)* } },
+        StructKind::Tuple => quote! { ( #(#fields,)* ); },
+    };
+
+    Ok(quote! {
+        #[doc = #doc]
+        #[derive(Debug, #derive_deserialize)]
+        #vis struct #incoming_ident #ty_gen #struct_def
+
+        impl #original_impl_gen ::ruma_api::Outgoing for #original_ident #original_ty_gen {
+            type Incoming = #incoming_ident #impl_gen;
+        }
+    })
+}
+
+fn no_deserialize_in_attrs(attrs: &[Attribute]) -> bool {
+    for attr in attrs {
+        match &attr.path {
+            Path { leading_colon: None, segments }
+                if segments.len() == 1 && segments[0].ident == "incoming_no_deserialize" =>
```

Comment:
You probably just kept / copied old code here, right? Seems like this is just `is_ident("incoming_no_deserialize")`.